src/wmtk/operations/tri_mesh/VertexAttributesUpdateBase.cpp

```diff
@@ -0,0 +1,65 @@
+#include "VertexAttributesUpdateBase.hpp"
+
+#include <wmtk/SimplicialComplex.hpp>
+#include <wmtk/TriMesh.hpp>
+
+namespace wmtk::operations {
+void OperationSettings<tri_mesh::VertexAttributesUpdateBase>::initialize_invariants(
+    const TriMesh& m)
+{
+    // outdated + is valid tuple
+    invariants = basic_invariant_collection(m);
+}
+
+namespace tri_mesh {
+VertexAttributesUpdateBase::VertexAttributesUpdateBase(
+    Mesh& m,
+    const Tuple& t,
+    const OperationSettings<VertexAttributesUpdateBase>& settings)
+    : TriMeshOperation(m)
+    , TupleOperation(settings.invariants, t)
+    , m_settings{settings}
+{}
+
+std::string VertexAttributesUpdateBase::name() const
+{
+    return "tri_mesh_vertex_attributes_update";
+}
+
+const Tuple& VertexAttributesUpdateBase::return_tuple() const
+{
+    return m_output_tuple;
+}
+
+bool VertexAttributesUpdateBase::before() const
+{
+    if (!mesh().is_valid_slow(input_tuple())) {
```

Comment:
default invariants should do the valid check for you - dont need to overwrite this at all